Rack: Accept generic X-Request-Start header

JamesHarker opened the pull request:
This patch makes the `t=` portion of the X-Request-Start header optional so that the request queuing time can be supported on platforms like Heroku.

marcotc commented on lib/ddtrace/contrib/rack/request_queue.rb:
```diff
@@ -13,9 +13,7 @@ def get_request_start(env, now = Time.now.utc)
           return unless header
 
           # nginx header is in the format "t=1512379167.574"
```
Could you update the comment to mention the format Heroku uses?

JamesHarker commented:
> I noticed that [Heroku's documentation](https://devcenter.heroku.com/articles/http-routing#heroku-headers) mentions that time is represented in milliseconds.
> 
> Our current implementation calls `Time.at(time_value)`, which expects time in seconds, which is what `nginx` uses.
> 
> To handle Heroku headers correctly, a conversion has to be made here.

Apologies, completely overlooked that. I'll update the PR 👍

JamesHarker commented:
@marcotc I've updated the PR as requested. I've also added support for Apache as apparently the default there is something different again 🙄 

I've added some more tests but I'm not sure if this if this is OTT? Happy to amend if needs be.

I think I've also forked (and have the PR merging into) the wrong branch - would you like me to change this?

marcotc commented on lib/ddtrace/contrib/rack/request_queue.rb:
```diff
@@ -12,14 +12,14 @@ def get_request_start(env, now = Time.now.utc)
 
           # Return nil if the time is clearly invalid
-          time_value = time_string.to_f
+          time_value = "#{time_string[0, 10]}.#{time_string[10, 13]}".to_f
```
This is quite clever!
One thing I suggest is reducing the number of fractional digits to 6, as that's the highest precision we currently expect.
But I'm curious, why you chose 13 fractional digits?

marcotc commented on lib/ddtrace/contrib/rack/request_queue.rb:
```diff
@@ -12,14 +12,14 @@ def get_request_start(env, now = Time.now.utc)
-          # TODO: this should be generic enough to work with any
-          # frontend web server or load balancer
-          time_string = header.split('t=')[1]
+          # nginx header is milliseconds in the format "t=1512379167.574"
```
`nginx` header is actually in seconds.

marcotc commented on lib/ddtrace/contrib/rack/request_queue.rb:
```diff
@@ -12,14 +12,14 @@ def get_request_start(env, now = Time.now.utc)
+          # nginx header is milliseconds in the format "t=1512379167.574"
+          # apache header is microseconds in the format "t=1570633834463123"
+          # heorku header is milliseconds in the format "1570634024294"
+          time_string = header.to_s.gsub(/t=|\./, '')
```
I was curious if this was the best way to remove all non-numeric characters from a string, so I [ran a benchmark](https://gist.github.com/marcotc/09db9af991b0ad3c435573836f86221e).

`String#delete` seems to be the fastest way, also using the least amount of memory, so I suggest this change:

```suggestion
          time_string = header.to_s.delete('^0-9')
```

marcotc commented on lib/ddtrace/contrib/rack/request_queue.rb:
```diff
@@ -12,14 +12,14 @@ def get_request_start(env, now = Time.now.utc)
+          time_value = "#{time_string[0, 10]}.#{time_string[10, 13]}".to_f
           return if time_value.zero?
 
           # return the request_start only if it's lesser than
```
Give the new code, could you update the sanity check that happens in the next few lines to reflect the new valid bounds.

I believe the new valid bounds are:
```ruby
> Time.at(9999999999)
=> 2286-11-20 18:46:39 +0100
> Time.at(1000000000)
=> 2001-09-09 03:46:40 +0200
```

You can return `nil` for numbers outside of this range.

JamesHarker replied:
So the max time bound isn't possible using the way I'm currently parsing the string because the decimal place is put back in the wrong place. For example consider a header of `t=10000000000`

```ruby
# Instead of it calculating the correct time (which should be invalid)
Time.at(10_000_000_000)
=>2286-11-20 17:46:40 +0000
# it actually becomes 1_000_000_000.0
Time.at(1_000_000_000.0)
=> 2001-09-09 01:46:40 +0000 # Which is now a valid time (just!)
```

I can make the code a slightly stricter and check for length to figure out if the timestamp is in seconds, milliseconds or microseconds using something like this:

```ruby
divisor = case time_string.length
when 16 # microseconds
  1_000_000
when 13 # milliseconds
  1_000
else
  1
end
time_value = time_string.to_f / divisor
```

However, this [test](https://github.com/DataDog/dd-trace-rb/blob/18300bd2c645c213c22ade2fd66b5bc6151be450/spec/ddtrace/contrib/rack/distributed_spec.rb#L103) then fails about ~50% of the time because precision of the time is being lost as a result of first converting the time it into a float and then into a string 😱 .

We can fix the test by retaining the precision using something like

```ruby
"%10.6f" % Time.now.to_f
 => "1570724109.724110"
```

However this will mean that times with decimal precision will have to have either 3 or 6 decimal places or the time will be rejected. I think this is ok because most web servers (well at least the 3 we're testing for here) will pass a correct timestamp. Additionally this is similar to what [other](https://github.com/newrelic/rpm/blob/17734c85a9251ede14fbee517f0d264d99116c7e/lib/new_relic/agent/instrumentation/queue_time.rb#L59-L69) [libs](https://github.com/librato/librato-rack/blob/4c5cd0cefec193acb5845199a66997dca2f2fe39/lib/librato/rack.rb#L150-L161) do but wanted to check before making the change?

Alternatively we can go back to the original solution and remove the upper bounds restriction. Yes, times from 2286 onwards will be incorrectly parsed but it's possibly something you're ok to live with in exchange for the reduced complexity?

marcotc replied:
> times with decimal precision will have to have either 3 or 6 decimal places or the time will be rejected

I think this solution could cause us problems for headers with decimal, as I can foresee them dropping least significant zeroes.

> times from 2286 onwards will be incorrectly parsed but it's possibly something you're ok to live with in exchange for the reduced complexity

This limitation seems reasonable to me.

I'm also not opposed to a solution similar to [this implementation](https://github.com/newrelic/rpm/blob/17734c85a9251ede14fbee517f0d264d99116c7e/lib/new_relic/agent/instrumentation/queue_time.rb#L59-L69). (In case you choose this one, skip boxing the timestamp into a `Time` object for checking bounds. That check can be done against the "raw" `Numeric` instance instead.)

My bias is towards correctness first, but maintainability is a close second: a less complex solution is favorable, given the limitations are reasonable.

Ultimately, given the extent of discussions we had in this PR so far, I'll trust your judgement to choose the implementation you think is best here.

marcotc replied:
👍

marcotc commented on lib/ddtrace/contrib/rack/request_queue.rb:
```diff
@@ -5,22 +5,23 @@ module Rack
-          time_string = header.split('t=')[1]
+          # nginx header is seconds in the format "t=1512379167.574"
+          # apache header is microseconds in the format "t=1570633834463123"
+          # heorku header is milliseconds in the format "1570634024294"
```
```suggestion
          # heroku header is milliseconds in the format "1570634024294"
```

marcotc commented:
Only a few housekeeping things left to do: address Rubocop issues and rebase with master.

JamesHarker commented:
> @JamesHarker The fix to the MongoDB test failure you are seeing in your build is in `master` branch. Sync with master and you are good!

Master merged back in so I think we're good go? 👍
